### ISSUE TITLE
yarn.lock should not include base registry(`http://registry.npmjs.org`)

### ISSUE THREAD

**Issue by doxiaodong** [yarnpkg/yarn#3330]
**Do you want to request a *feature* or report a *bug*?**

feature
**What is the current behavior?**
yarn.lock include registry
**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**
```
lodash@4.17.4:
  version "4.17.4"
  resolved "http://registry.npm.taobao.org/lodash/download/lodash-4.17.4.tgz#78203a4d1c328ae1d86dca6460e369b57f4055ae"
```
should be
```
...
resolved "/lodash/download/lodash-4.17.4.tgz#78203a4d1c328ae1d86dca6460e369b57f4055ae"
...
```
 **Reason**
Most developers in china will use `--registry=https://registry.npm.taobao.org
`, and then, yarn.lock will be 
```
...
resolved "http://registry.npm.taobao.org/lodash/download/lodash-4.17.4.tgz#78203a4d1c328ae1d86dca6460e369b57f4055ae"
...
```
But for travis-ci, circleci, it seem to be slowly.
So, I think the regiistry should not be in yarn.lock

**Comment by bestander** [comment]
Yeah, that sounds like a useful feature.
@arcanis what do you think?

**Comment by arcanis** [comment]
Yup, completely agree. Ideally, I think the `resolved` field should be removed (because we don't care where the tarball come from, as long as it has the right hash), and replaced by a single `hash` field. The url can then be reconstructed at runtime when reading the lockfile.

**Comment by bestander** [comment]
@doxiaodong want to send an RFC?

We'll need to discuss backwards compatibility, this feature being an opt-in, defaults etc

**Comment by doxiaodong** [comment]
@bestander ， any template here? I don't know what and how to do that.

**Comment by bestander** [comment]
Sure, here are a couple of good examples https://github.com/yarnpkg/rfcs/pull/40/files https://github.com/yarnpkg/rfcs/pull/34/files

**Comment by doxiaodong** [comment]
I have sent a pr https://github.com/yarnpkg/rfcs/pull/64

**Comment by csvan** [comment]
+1 for this. In our case (which I think is shared by many) we have one repository in our local dev environment, and another in the actual build environment. Currently, tracking the lock file with git causes it to contain references to the dev registry which in turn leads to builds failing. The only solution we have found so far is to not track the lock file at all, which incurs a performance hit during builds.

**Comment by ckannon** [comment]
+1 for this. We use a local registry and would like it to ignore the one in the yarn.lock in favor of our local.

**Comment by el-davo** [comment]
+1 for this. In our situation we are using jenkins and sinopia in a kubernetes environment. In the kubernetes network sinopia has a different url to the external network url. It would be handy if we could switch between the 2 easily as it might be faster to use the internal kubernetes network url, while allowing developers to use the external registry url as they wont have access to the internal url

**Comment by vernak2539** [comment]
This would be nice to have. We have a mirror used outside our Build Environment that has one URL. Inside the Build Environment it has a different URL, and the original registry used to generate the lock file is not available. 

We've noticed that passing the `--registry` flag doesn't overwrite from where the packages are fetched. Would be nice to have the lock file be registry agnostic

**Comment by bestander** [comment]
Suggest a simple backwards compatible RFC and send a PR, we'll get it sorted then.
Discussion in https://github.com/yarnpkg/rfcs/pull/64 got derailed and suggest bigger changes than the team is eager to take.

**Comment by vernak2539** [comment]
@bestander I wrote one [here](https://github.com/yarnpkg/rfcs/pull/84). Not sure if it's enough, but it's a start

**Comment by aikar** [comment]
@bestander when you say backwards compat RFC, do you mean newer yarn can still load older yarnrc, or that older yarns should still be able to load the newer yarnrc?

**Comment by bestander** [comment]
@aikar, conceptually I agree.
Although a major change in yarn.lock would need a detailed RFC with consideration of how breaking the change will be.

When I talk about backwards compatibility I mean that yarn.lock generated by a previous version of Yarn won't be compatible by the next version that strips the URLs and vice versa.
We'll need to think through how teams with people using multiple versions of Yarn can work.
It is doable but requires a responsible approach.

So far I am fine to settle on a minimal change to unblock people using npm mirrors.

**Comment by Maxxxz** [comment]
Met this question today ToT

**Comment by ashwinr** [comment]
This has sadly become a blocker for us. Npm works correctly using an `integrity` field in its lock file (in addition to a `resolved` field similar to yarn). Here's a solution that also preserves yarn backcompat: Introduce a _new_ field called `hash` that's a sibling of `resolved`. Latest yarn would add this field wherever it's missing for each package entry in the lock file, when `yarn install` is run. When resolving a package, if `hash` is missing in any entry in the lock file, yarn falls back to the `resolved` field, which will keep backcompat with old lock files. Thoughts?

**Comment by bestander** [comment]
Here is the RFC we have agreed on, it should solve the title issue and is waiting for a champion to implement it: https://github.com/yarnpkg/rfcs/pull/84/files.

It is quite easy to implement it so please go ahead and send a PR.

As for talks about integrity fields we have one here: https://github.com/yarnpkg/yarn/issues/816

**Comment by crazyyi** [comment]
I hope this can be done soon. This is really annoying.

**Comment by LinkMJB** [comment]
Has there been any progress on this? Not being able to utilize a proxy or mirror without circumventing the source controlled `yarn.lock` is really frustrating. Especially when issues like this contribute: 

https://status.npmjs.org/incidents/vrg1s7lbg887
```
An issue with all logged-out site users being put into the same rate-limiting bucket has been fixed.
```

https://status.npmjs.org/incidents/rmv88ktdgkkr
```
Some users are reporting timeouts on their package installations. We're currently investigating this.
```

**Comment by LinkMJB** [comment]
Here's a quick workaround I slapped together to circumvent this behavior until a **REAL** patch is implemented. You should be able to execute this as a pre-step in CI jobs to update every instance of `yarn.lock` to use what was established as the registry in the `.yarnrc`:

```
find . -name yarn.lock -exec sed -i "s#https://registry.yarnpkg.com#`awk '$1 == "registry" { gsub(/"/,""); print $NF}' /root/.yarnrc`#g" {} \;
```

**Comment by LinkMJB** [comment]
Wow, that was impeccable timing 🤣 

https://status.yarnpkg.com/incident/14/

**Comment by arcanis** [comment]
Opened a new issue at #5892 to put this feature request inside the v2 high level goals 🙂

**Comment by BYK** [comment]
@arcanis shall we close this one in favor of #5892?

**Comment by lcw0622** [comment]
So this feature still under development?
It is really important for deploy in production environment when company has own internal registry.

**Comment by coppyC** [comment]
you are right

**Comment by arcanis** [comment]
Closing this issue; we already have #5892.

This will be part of the v2, but probably not the v1.

**Comment by icyerasor** [comment]
What about at least logging that there are inconsistencies between the configured registry and the cached URLs in the lock?
Would probably save lots of dev hrs.

**Comment by kopax-polyconseil** [comment]
1. The checksum of the yarn lock change if you edit it, that's one problem
2. The git repo not being up to date after editing the yarn lock, a second problem. 
3. A workaround would be to allow to specifiy the lock file then create a 2nd one with the first one using the other repo, that's not possible
4. A setting stronger than `--repository` that could take precedence over the lock file, also not possible.

Any plan to implement a solution within yarn after all those years ?

**Comment by arcanis** [comment]
> Any plan to implement a solution within yarn after all those years ?

It got fixed about three years and two major versions ago.

**Comment by hmatfarafan** [comment]
> > Any plan to implement a solution within yarn after all those years ?
> 
> It got fixed about three years and two major versions ago.

My current problem  is that it seems all official [node docker images](https://hub.docker.com/_/node) include version 1.x.x of yarn and it is a ban of our CI processes to use latest versions of yarn.

**Comment by chadlwilson** [comment]
> it is a ban of our CI processes to use latest versions of yarn.

Well, no-one can help you with this. You either need to stop using Yarn, or workaround your policy to upgrade.

When used with [corepack](https://yarnpkg.com/corepack), [making your build use Yarn 2+](https://yarnpkg.com/migration/guide#migration-steps) with external node docker images is pretty trivial.

Yarn 2+ can essentially be treated as a build-time dependency like any other (not something pre-installed) and can be source controlled/locked alongside your code etc so doesn't really "affect" your CI environment except if you are doing something special to mount the Yarn cache for re-use between builds. In other words, Yarn 2+ does need to be, and is canonically never pre-installed on something like a container image. Rational explained on the linked docs.

**Comment by hmatfarafan** [comment]
@chadlwilson 
Ok,
But if I want to be more clear, we use yarn.lock to be sure about keeping dependencies of the project intact. And then to speed up CI process, we prefer to set a hosted local registry and it cannot be done in yarn 1.x.
You say we can add Yarn 2+ as a build-time dependency. It means every CI run leads to run the process of Yarn 2+ installation again, and it is time consuming and makes the CI process slow down.

May keeping Yarn 1.x and downloading artifacts from original registry every time be simpler :-) .

**Comment by chadlwilson** [comment]
That’s now how it works or has to work (read the migration guide and/or my comment more closely around source controlling yarn itself) but this isn’t the right place for that discussion.
